src/util/pl-X10.1
.TH PL\-X11 1 BRL-CAD
.\"                       P L - X 1 0 . 1
.\" BRL-CAD
.\"
.\" Copyright (c) 2005-2007 United States Government as represented by
.\" the U.S. Army Research Laboratory.
.\"
.\" Redistribution and use in source (Docbook format) and 'compiled'
.\" forms (PDF, PostScript, HTML, RTF, etc), with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code (Docbook format) must retain the
.\" above copyright notice, this list of conditions and the following
.\" disclaimer.
.\"
.\" 2. Redistributions in compiled form (transformed to other DTDs,
.\" converted to PDF, PostScript, HTML, RTF, and other formats) must
.\" reproduce the above copyright notice, this list of conditions and
.\" the following disclaimer in the documentation and/or other
.\" materials provided with the distribution.
.\"
.\" 3. The name of the author may not be used to endorse or promote
.\" products derived from this documentation without specific prior
.\" written permission.
.\"
.\" THIS DOCUMENTATION IS PROVIDED BY THE AUTHOR AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
.\" OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS DOCUMENTATION, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\".\".\"
.SH NAME
pl\-X10 \- display UnixPlot in an X Window (X10)
.SH SYNOPSIS
.B pl-X10
.RB [ \-v ]
< unix_plot
.SH DESCRIPTION
.I Pl-X10
is a version for the X10R4 release of the X Window System.
It is provided as a last resort for those people who have
not upgraded to X11 yet.  It will be removed in a future
release.
.PP
The
.B \-v
verbose option causes descriptions of the plot commands to be
printed.
.SH BUGS
This is just the bare bones of a display program at the time
of this writing.  It needs the ability to rotate, translate, zoom, etc.
.SH "SEE ALSO"
pl-X(1), pl-fb(1), pl-sgi(1), pl-pl(1), plot3(5).
.SH AUTHORS
Phillip Dykstra
.SH SOURCE
SECAD/VLD Computing Consortium, Bldg 394
.br
The U. S. Army Ballistic Research Laboratory
.br
Aberdeen Proving Ground, Maryland  21005
.SH "BUG REPORTS"
Reports of bugs or problems should be submitted via electronic
mail to <devs@brlcad.org>.


NEWS
						 -*- coding: utf-8 -*-
BRL-CAD Release Notes
=====================

Included below is the latest news pertaining to BRL-CAD releases.  The
history or BRL-CAD prior to release 5.0 is covered in doc/history.txt,
see that file and the CVS log for more historic information.  The
entire history of the project is available through CVS.

----------------------------------------------------------------------
--- 2006-12-XX  Release 7.9.0                                      ---
----------------------------------------------------------------------

With this release marks several significant changes to BRL-CAD's
licensing in order to simplify the distribution terms and increase
overall flexibility of use.  All application front-end code in BRL-CAD
is converted from the GNU General Public License (GPLv2) to the less
restrictive GNU Lesser General Public License (LGPLv2.1).

For documents not already in the public domain, BRL-CAD's default
documentation license terms are also changed from the dual-licensed
GFDL/GPL terms to the BSD Documentation License (BDL).  Originally
derived from the FreeBSD Documentation License, the BDL is effectively
the BSD License with language as it applies specifically to
documentation.  The terms of the BDL are just as flexible allowing
practically any use, whether it be commercial or private, so long as
the authors are appropriately credited.

As a collective work, BRL-CAD is being distributed under the terms of
the LGPL.  Individual portions of BRL-CAD, save for a few 3rd party
codes, effectively fall into one of the following three license
categories: 1) LGPL, 2) BSD/BDL, or 3) are in the public domain.  See
the terms of each individual license as well as BRL-CAD's COPYING file
for more details.

* updated bundled zlib to version 1.2.3 - Sean Morrison
* updated bundled libpng to version 1.2.16 - Sean Morrison
* g_qa manual page documents the -t tolerance option - Lee Butler
* dxf-g turns 2D entities into sketches instead of nmg - John Anderson
* removed advertising clause from BSD code - Sean Morrison
* fix index bug, allow material ID of zero in rtweight - Karel Kulhavy
* fixed multiple frame render bug in raytracers - Karel Kulhavy
* mged help command now shows help for all args listed - Sean Morrison


----------------------------------------------------------------------
--- 2006-10-16  Release 7.8.4                                      ---
----------------------------------------------------------------------

* support for handling larger hierarchies in archer - Doug Howard
* added distance measurement tool to archer - Bob Parker
* added component selection capabilites to archer - Bob Parker
* new script to extract gforge task items - Lee Butler, Sean Morrison
* substantial additions and mods to source documentation - Lee Butler
* added scripts to facetize all regions in a database - Lee Butler
* added script to expand combinations to multiple objects - Lee Butler
* prevent infinite loop on classic mged attach prompt - Sean Morrison
* ctrl-d on empty shell line now quits mged - Sean Morrison
* added archer geometry/view loading via command-line arg - Bob Parker
* mged tree command now supports -o/-i, matches docs - John Anderson
* improved beginning-of-line input processing in stl-g - John Anderson
* improved end-of-line input processing in dxf-g - John Anderson
* support for longer Fastgen object names in fast4-g - Sean Morrison
* improved 'ted' mged command text editor support - Sean Morrison
* support mged EDITOR environment override on Windows - Sean Morrison
* fixed mged crash on P binding when not in edit mode - Sean Morrison
* improved FreeBSD 64bit support, fixes AMD64 crash - Erik Greenwald
* allow unlimited light samples during ray tracing - Sean Morrison
* new script for computing BRL-CAD file line counts - Sean Morrison
* fixed mged extrude of empty sketch display crash - Sean Morrison
* re-enabled png-ipu, pix-ipu, ipustat, ipuscan - Sean Morrison
* updated archer splash screen - Bob Parker
* archer enhancements for displaying view parameters - Bob Parker
* improved mged grid zooming using pixel size - John Anderson
* fixed mged hang caused by extreme grid zooming bug - John Anderson
* ported jove to SGI Altix platform, fixed crash bug - Sean Morrison
* increase default mged scrollback to 10k lines - Sean Morrison
* improved support for Mac "delete" keys in mged - Sean Morrison
* mged vi-mode command edit history searching - James (swcto)
* bigger, better vi command line editing in mged - James (swcto)
* rt command script 'ae' no longer resets view scale - Sean Morrison
* fixed asc-nmg manual page usage examples - Sean Morrison


----------------------------------------------------------------------
--- 2006-06-20  Release 7.8.2                                      ---
----------------------------------------------------------------------

This release includes a new example geometry model, a "Toy Barbie
Jeep" (see note below regarding trademarks) that was modeled in 2005
over the course of a day by a small group of students attending a free
hands-on BRL-CAD training class at SURVICE Engineering, Inc.  The
class involved working in groups to simulate a real-world
time-constrained modeling experience taking measurements of a real toy
vehicle, considering what aspects of the vehicle are important for
analytic purposes, deciding how to best represent complex geometric
shapes, modeling the parts of the vehicle in MGED, assembling the
components together, and running geometric processing and rendering
tools on the final model.

The DXF importer received considerable updates this iteration, with
John Anderson providing additional support for a variety of 2D DXF
entities.

* fixed asc-nmg bug that caused a crash on exit - John Anderson
* fixed minor mged memory leak - Jim Monte
* added archer version info to bottom of about dialog - Bob Parker
* added a beta vertex array exporter (g-var) - Prasad Silva

src/util/pl-X10.c
/*                        P L - X 1 0 . C
 * BRL-CAD
 *
 * Copyright (c) 1988-2007 United States Government as represented by
 * the U.S. Army Research Laboratory.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this file; see the file named COPYING for more
 * information.
 */
/** @file pl-X10.c
 *
 *  Display plot3(5) on an X Window System display (X10R4)
 *
 *  Author -
 *	Phillip Dykstra
 *
 *  Source -
 *	SECAD/VLD Computing Consortium, Bldg 394
 *	The U. S. Army Ballistic Research Laboratory
 *	Aberdeen Proving Ground, Maryland  21005-5066
 */
#ifndef lint
static const char RCSid[] = "@(#)$Header$ (BRL)";
#endif

#include "common.h"

#include <stdio.h>
#include <X/Xlib.h>

#include "machine.h"

#define	TBAD	0	/* no such command */
#define TNONE	1	/* no arguments */
#define TSHORT	2	/* Vax 16-bit short */
#define	TIEEE	3	/* IEEE 64-bit floating */
#define	TCHAR	4	/* unsigned chars */
#define	TSTRING	5	/* linefeed terminated string */

struct uplot {
	int	targ;	/* type of args */
	int	narg;	/* number or args */
	char	*desc;	/* description */
};
struct uplot uerror = { 0, 0, 0 };
struct uplot letters[] = {
/*A*/	{ 0, 0, 0 },
/*B*/	{ 0, 0, 0 },
/*C*/	{ TCHAR, 3, "color" },
/*D*/	{ 0, 0, 0 },
/*E*/	{ 0, 0, 0 },
/*F*/	{ TNONE, 0, "flush" },
/*G*/	{ 0, 0, 0 },
/*H*/	{ 0, 0, 0 },
/*I*/	{ 0, 0, 0 },
/*J*/	{ 0, 0, 0 },
/*K*/	{ 0, 0, 0 },
/*L*/	{ TSHORT, 6, "3line" },
/*M*/	{ TSHORT, 3, "3move" },
/*N*/	{ TSHORT, 3, "3cont" },
/*O*/	{ TIEEE, 3, "d_3move" },
/*P*/	{ TSHORT, 3, "3point" },
/*Q*/	{ TIEEE, 3, "d_3cont" },
/*R*/	{ 0, 0, 0 },
/*S*/	{ TSHORT, 6, "3space" },
/*T*/	{ 0, 0, 0 },
/*U*/	{ 0, 0, 0 },
/*V*/	{ TIEEE, 6, "d_3line" },
/*W*/	{ TIEEE, 6, "d_3space" },
/*X*/	{ TIEEE, 3, "d_3point" },
/*Y*/	{ 0, 0, 0 },
/*Z*/	{ 0, 0, 0 },
/*[*/	{ 0, 0, 0 },
/*\*/	{ 0, 0, 0 },
/*]*/	{ 0, 0, 0 },
/*^*/	{ 0, 0, 0 },
/*_*/	{ 0, 0, 0 },
/*`*/	{ 0, 0, 0 },
/*a*/	{ TSHORT, 6, "arc" },
/*b*/	{ 0, 0, 0 },
/*c*/	{ TSHORT, 3, "circle" },
/*d*/	{ 0, 0, 0 },
/*e*/	{ TNONE, 0, "erase" },
/*f*/	{ TSTRING, 1, "linmod" },
/*g*/	{ 0, 0, 0 },
/*h*/	{ 0, 0, 0 },
/*i*/	{ TIEEE, 3, "d_circle" },
/*j*/	{ 0, 0, 0 },
/*k*/	{ 0, 0, 0 },
/*l*/	{ TSHORT, 4, "line" },
/*m*/	{ TSHORT, 2, "move" },
/*n*/	{ TSHORT, 2, "cont" },
/*o*/	{ TIEEE, 2, "d_move" },
/*p*/	{ TSHORT, 2, "point" },
/*q*/	{ TIEEE, 2, "d_cont" },
/*r*/	{ TIEEE, 6, "d_arc" },
/*s*/	{ TSHORT, 4, "space" },
/*t*/	{ TSTRING, 1, "label" },
/*u*/	{ 0, 0, 0 },
/*v*/	{ TIEEE, 4, "d_line" },
/*w*/	{ TIEEE, 4, "d_space" },
/*x*/	{ TIEEE, 2, "d_point" },
/*y*/	{ 0, 0, 0 },
/*z*/	{ 0, 0, 0 }
};

double	getieee();
int	verbose;
double	cx, cy, cz;		/* current x, y, z, point */
double	arg[6];			/* parsed plot command arguments */
double	sp[6];			/* space command */
char	strarg[512];		/* string buffer */

static char usage[] = "\
Usage: pl-X10 [-v] < unix_plot\n";

Window	w;
FontInfo *font;
WindowInfo winfo;

int
main( argc, argv )
int	argc;
char	**argv;
{
	register int	c;
	struct	uplot *up;
	int	i;

	while( argc > 1 ) {
		if( strcmp(argv[1], "-v") == 0 ) {
			verbose++;
		} else
			break;

		argc--;
		argv++;
	}
	if( isatty(fileno(stdin)) ) {
		fprintf( stderr, usage );
		exit( 1 );
	}
	xsetup( argc, argv );

	while( (c = getchar()) != EOF ) {
		/* look it up */
		if( c < 'A' || c > 'z' ) {
			up = &uerror;
		} else {
			up = &letters[ c - 'A' ];
		}

		if( up->targ == TBAD ) {
			fprintf( stderr, "Bad command '%c' (0x%02x)\n", c, c );
			continue;
		}

		if( up->narg > 0 )
			getargs( up );

		switch( c ) {
		case 's':
		case 'w':
			sp[0] = arg[0];
			sp[1] = arg[1];
			sp[2] = 0;
			sp[3] = arg[2];
			sp[4] = arg[3];
			sp[5] = 0;
			break;
		case 'S':
		case 'W':
			sp[0] = arg[0];
			sp[1] = arg[1];
			sp[2] = arg[2];
			sp[3] = arg[3];
			sp[4] = arg[4];
			sp[5] = arg[5];
			break;
		case 'm':
		case 'o':
			cx = arg[0];
			cy = arg[1];
			cz = 0;
			break;
		case 'M':
		case 'O':
			cx = arg[0];
			cy = arg[1];
			cz = arg[2];
			break;
		case 'n':
		case 'q':
			draw( cx, cy, cz, arg[0], arg[1], 0.0 );
			break;
		case 'N':
		case 'Q':
			draw( cx, cy, cz, arg[0], arg[1], arg[2] );
			break;
		case 'l':
		case 'v':
			draw( arg[0], arg[1], 0.0, arg[2], arg[3], 0.0 );
			break;
		case 'L':
		case 'V':
			draw( arg[0], arg[1], arg[2], arg[3], arg[4], arg[5] );
			break;
		case 'p':
		case 'x':
			draw( arg[0], arg[1], 0.0, arg[0], arg[1], 0.0 );
			break;
		case 'P':
		case 'X':
			draw( arg[0], arg[1], arg[2], arg[0], arg[1], arg[2] );
			break;
		case 't':
			label( cx, cy, strarg );
			break;
		case 'F':
			XFlush();
			break;
		case 'e':
			XClear( w );
			break;
		}

		if( verbose )
			printf( "%s\n", up->desc );
	}

	XFlush();
	sleep( 1 );
	return 0;
}

getargs( up )
struct uplot *up;
{
	int	i;

	for( i = 0; i < up->narg; i++ ) {
		switch( up->targ ){
			case TSHORT:
				arg[i] = getshort();
				break;
			case TIEEE:
				arg[i] = getieee();
				break;
			case TSTRING:
				getstring();
				break;
			case TCHAR:
				arg[i] = getchar();
				break;
			case TNONE:
			default:
				arg[i] = 0;	/* ? */
				break;
		}
	}
}

getstring()
{
	int	c;
	char	*cp;

	cp = strarg;
	while( (c = getchar()) != '\n' && c != EOF )
		*cp++ = c;
	*cp = 0;
}

getshort()
{
	register long	v, w;

	v = getchar();
	v |= (getchar()<<8);	/* order is important! */

	/* worry about sign extension - sigh */
	if( v <= 0x7FFF )  return(v);
	w = -1;
	w &= ~0x7FFF;
	return( w | v );
}

double
getieee()
{
	char	in[8];
	double	d;

	fread( in, 8, 1, stdin );
	ntohd( &d, in, 1 );
	return	d;
}

draw( x1, y1, z1, x2, y2, z2 )
double	x1, y1, z1;	/* from point */
double	x2, y2, z2;	/* to point */
{
	int	sx1, sy1, sx2, sy2;

	sx1 = (x1 - sp[0]) / (sp[3] - sp[0]) * winfo.width;
	sy1 = winfo.height - (y1 - sp[1]) / (sp[4] - sp[1]) * winfo.height;
	sx2 = (x2 - sp[0]) / (sp[3] - sp[0]) * winfo.width;
	sy2 = winfo.height - (y2 - sp[1]) / (sp[4] - sp[1]) * winfo.height;

	XLine( w, sx1, sy1, sx2, sy2, 1, 1, BlackPixel, GXand, AllPlanes );

	cx = x2;
	cy = y2;
	cz = z2;
}

label( x, y, str )
double	x, y;
char	*str;
{
	int	sx, sy;

	sx = (x - sp[0]) / (sp[3] - sp[0]) * winfo.width;
	sy = winfo.height - (y - sp[1]) / (sp[4] - sp[1]) * winfo.height;
	sy -= font->height;	/* point is lower left of text */

	XText( w, sx, sy, str, strlen(str), font->id, BlackPixel, WhitePixel );
}

xsetup( argc, argv )
int	argc;
char	**argv;
{
	int	width, height;
	int	defwidth, defheight;
	char	def[80];	/* default size string */
	char	*geometry;
	OpaqueFrame	frame;
	char	hostname[80];
	char	display[80];
	char	*envp;

	width = height = 512;

	if( (envp = getenv("DISPLAY")) == NULL ) {
		/* Env not set, use local host */
		gethostname( hostname, 80 );
		sprintf( display, "%s:0", hostname );
		envp = display;
	}

	if( XOpenDisplay( envp ) == NULL ) {
		fprintf( stderr, "pl-X10: Can't open X display\n" );
		exit( 2 );
	}
	defwidth = width + 10;
	defheight = height + 10;
	(void)sprintf( def, "=%dx%d+300+300", defwidth, defheight );
	geometry = "";
	frame.background = WhitePixmap;
	frame.border = BlackPixmap;
	frame.bdrwidth = 1;
	w = XCreate( "Unix Plot", argv[0],
		geometry, def, &frame,
		10, 10 );		/* minimum size */
	if( w == 0 ) {
		fprintf( stderr, "pl-X10: Can't create window\n" );
		exit( 3 );
	}
	XMapWindow( w );
	XQueryWindow( w, &winfo );

	if( !(font = XOpenFont("vtsingle")) ) {
		fprintf( stderr, "pl-X10: Can't open font\n" );
		exit( 4 );
	}
}

/*
 * Local Variables:
 * mode: C
 * tab-width: 8
 * c-basic-offset: 4
 * indent-tabs-mode: t
 * End:
 * ex: shiftwidth=4 tabstop=8
 */
